// _vant@3.0.0-beta.5@vant/es/action-sheet/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createVNode, mergeProps } from "vue";
// Utils
import { createNamespace, pick } from '../utils'; // Components

import Icon from '../icon';
import Popup, { popupSharedProps } from '../popup';
import Loading from '../loading';

var _createNamespace = createNamespace('action-sheet'),
    createComponent = _createNamespace[0],
    bem = _createNamespace[1];

export default createComponent({
  props: _extends({}, popupSharedProps, {
    title: String,
    actions: Array,
    cancelText: String,
    description: String,
    closeOnPopstate: Boolean,
    closeOnClickAction: Boolean,
    round: {
      type: Boolean,
      default: true
    },
    closeable: {
      type: Boolean,
      default: true
    },
    closeIcon: {
      type: String,
      default: 'cross'
    },
    safeAreaInsetBottom: {
      type: Boolean,
      default: true
    }
  }),
  emits: ['select', 'cancel', 'update:show'],
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        emit = _ref.emit;
    var popupPropKeys = Object.keys(popupSharedProps);

    var onUpdateShow = function onUpdateShow(show) {
      emit('update:show', show);
    };

    var onCancel = function onCancel() {
      onUpdateShow(false);
      emit('cancel');
    };

    var renderHeader = function renderHeader() {
      if (props.title) {
        return createVNode("div", {
          "class": bem('header')
        }, [props.title, props.closeable && createVNode(Icon, {
          "name": props.closeIcon,
          "class": bem('close'),
          "onClick": onCancel
        }, null)]);
      }
    };

    var renderCancel = function renderCancel() {
      if (props.cancelText) {
        return [createVNode("div", {
          "class": bem('gap')
        }, null), createVNode("button", {
          "type": "button",
          "class": bem('cancel'),
          "onClick": onCancel
        }, [props.cancelText])];
      }
    };

    var renderOption = function renderOption(item, index) {
      var name = item.name,
          color = item.color,
          subname = item.subname,
          loading = item.loading,
          disabled = item.disabled,
          className = item.className;
      var Content = loading ? createVNode(Loading, {
        "class": bem('loading-icon')
      }, null) : [createVNode("span", {
        "class": bem('name')
      }, [name]), subname && createVNode("div", {
        "class": bem('subname')
      }, [subname])];

      var onClick = function onClick() {
        emit('select', item, index);

        if (props.closeOnClickAction) {
          onUpdateShow(false);
        }
      };

      return createVNode("button", {
        "type": "button",
        "style": {
          color: color
        },
        "class": [bem('item', {
          loading: loading,
          disabled: disabled
        }), className],
        "onClick": onClick
      }, [Content]);
    };

    var renderDescription = function renderDescription() {
      if (props.description || slots.description) {
        var content = slots.description ? slots.description() : props.description;
        return createVNode("div", {
          "class": bem('description')
        }, [content]);
      }
    };

    var renderOptions = function renderOptions() {
      if (props.actions) {
        return props.actions.map(renderOption);
      }
    };

    return function () {
      return createVNode(Popup, mergeProps({
        "class": bem(),
        "round": props.round,
        "position": "bottom"
      }, _extends({}, pick(props, popupPropKeys), {
        'onUpdate:show': onUpdateShow
      })), {
        default: function _default() {
          return [renderHeader(), renderDescription(), renderOptions(), slots.default && createVNode("div", {
            "class": bem('content')
          }, [slots.default()]), renderCancel()];
        }
      });
    };
  }
});
