// _vant@3.0.0-beta.5@vant/es/address-list/Item.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createVNode, resolveDirective } from "vue";
// Utils
import { createNamespace } from '../utils'; // Components

import Tag from '../tag';
import Icon from '../icon';
import Cell from '../cell';
import Radio from '../radio';

var _createNamespace = createNamespace('address-item'),
    createComponent = _createNamespace[0],
    bem = _createNamespace[1];

export default createComponent({
  props: {
    data: Object,
    disabled: Boolean,
    switchable: Boolean,
    defaultTagText: String
  },
  emits: ['edit', 'click', 'select'],
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        emit = _ref.emit;

    var onClick = function onClick() {
      if (props.switchable) {
        emit('select');
      }

      emit('click');
    };

    var renderRightIcon = function renderRightIcon() {
      return createVNode(Icon, {
        "name": "edit",
        "class": bem('edit'),
        "onClick": function onClick(event) {
          event.stopPropagation();
          emit('edit');
          emit('click');
        }
      }, null);
    };

    var renderTag = function renderTag() {
      if (props.data.isDefault && props.defaultTagText) {
        return createVNode(Tag, {
          "type": "danger",
          "round": true,
          "class": bem('tag')
        }, {
          default: function _default() {
            return [props.defaultTagText];
          }
        });
      }
    };

    var renderContent = function renderContent() {
      var data = props.data,
          disabled = props.disabled,
          switchable = props.switchable;
      var Info = [createVNode("div", {
        "class": bem('name')
      }, [data.name + " " + data.tel, renderTag()]), createVNode("div", {
        "class": bem('address')
      }, [data.address])];

      if (switchable && !disabled) {
        return createVNode(Radio, {
          "name": data.id,
          "iconSize": 18
        }, {
          default: function _default() {
            return [Info];
          }
        });
      }

      return Info;
    };

    return function () {
      var disabled = props.disabled;
      return createVNode("div", {
        "class": bem({
          disabled: disabled
        }),
        "onClick": onClick
      }, [createVNode(Cell, {
        "border": false,
        "valueClass": bem('value')
      }, {
        default: renderContent,
        'right-icon': renderRightIcon
      }), slots.bottom == null ? void 0 : slots.bottom(_extends({}, props.data, {
        disabled: disabled
      }))]);
    };
  }
});
