// _vant@3.0.0-beta.5@vant/es/overlay/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { Transition, vShow, createVNode, withDirectives } from "vue";
import { noop, isDef, preventDefault, createNamespace } from '../utils';
import { useLazyRender } from '../composition/use-lazy-render';

var _createNamespace = createNamespace('overlay'),
    createComponent = _createNamespace[0],
    bem = _createNamespace[1];

export default createComponent({
  props: {
    show: Boolean,
    zIndex: [Number, String],
    duration: [Number, String],
    className: null,
    customStyle: Object,
    lockScroll: {
      type: Boolean,
      default: true
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
    var lazyRender = useLazyRender(function () {
      return props.show;
    });

    var preventTouchMove = function preventTouchMove(event) {
      preventDefault(event, true);
    };

    var renderOverlay = lazyRender(function () {
      var style = _extends({
        zIndex: props.zIndex !== undefined ? +props.zIndex : undefined
      }, props.customStyle);

      if (isDef(props.duration)) {
        style.animationDuration = props.duration + "s";
      }

      return withDirectives(createVNode("div", {
        "style": style,
        "class": [bem(), props.className],
        "onTouchmove": props.lockScroll ? preventTouchMove : noop
      }, [slots.default == null ? void 0 : slots.default()]), [[vShow, props.show]]);
    });
    return function () {
      return createVNode(Transition, {
        "name": "van-fade"
      }, {
        default: function _default() {
          return [renderOverlay()];
        }
      });
    };
  }
});
